syntax = "proto3";
import "google/protobuf/empty.proto";
package factorio_server;

service ServerManager {
  rpc GetManagerStatus (google.protobuf.Empty) returns (ManagerStat);
  rpc GetAllSaveName (google.protobuf.Empty) returns (SaveNameList);
  rpc GetStatByName (SaveName) returns (SaveStat);
  rpc StopServer (google.protobuf.Empty) returns (Status);
  rpc StartServerByName (ServerOptions) returns (Status);
  rpc RestartServer (ServerOptions) returns (Status);
  rpc InGameCommand (Command) returns (Status);
  rpc WaitForUpdates (UpdateInquiry) returns (GameUpdates);
  rpc GetOutputStreams (google.protobuf.Empty) returns (OutputStreams);
}

message ManagerStat {
  string welcome = 1;
  bool running = 2;
}

message SaveNameList {
  repeated SaveName save_name = 1;
}

message SaveName {
  string name = 1;
}

// json encoded metadata of the saved file
message SaveStat {
  string stat_json = 1;
}

// server starting options
message ServerOptions {
  optional SaveName save_name = 1;
  repeated string extra_args = 2;
}

// start/stop status
message Status {
  int32 code = 1;
  string message = 2;
}

message Command {
  string cmd = 1;
}

message UpdateInquiry {
  optional int32 from_offset = 1;
}

message GameUpdates {
  int32 latest_offset = 1;
  repeated bytes updates = 2;
}

message OutputStreams {
  bytes stdout = 1;
  bytes stderr = 2;
}
